[receiver/sqlquery] Add Oracle DB support

crobert-1 opened the pull request:
**Description:** <Describe what has changed.>
This adds support for an Oracle DB driver to be used with the sqlquery receiver. This also includes testing to make
sure the receiver can connect and query metrics from a docker container running an Oracle DB. The README has
been updated, and a sample config has also been provided. The integration test suite has been updated to allow the
docker image for the Oracle DB to be created pre-test, and deleted post-test.

**Link to tracking Issue:** <Issue number if applicable>
[12137](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/12137)

**Testing:** <Describe what testing was performed and which tests were added.>
Integration test added

**Documentation:** <Describe the documentation added.>
README updated and sample config file provided.

crobert-1 committed:
[receiver/sqlquery] Add Oracle DB support

This adds support for an Oracle DB driver to be used with
the sqlquery receiver. This also includes testing to make
sure the receiver can connect and query metrics from a
docker container running an Oracle DB. The README has
been updated, and a sample config has also been provided.
The integration test suite has been updated to allow the
docker image for the Oracle DB to be created pre-test,
and deleted post-test.

pmcollins commented on receiver/sqlqueryreceiver/README.md:
```diff
@@ -22,8 +23,8 @@ A _query_ consists of a sql statement and one or more _metrics_, where each metr
 * `metric_name`(required): the name assigned to the OTel metric.
-* `value_column`(required): the column name in the returned dataset used to set the value of the metric's datapoint. The column's values must be of an integer type.
-* `attribute_columns`(optional): a list of column names in the returned dataset used to set attibutes on the datapoint.
+* `value_column`(required): the column name in the returned dataset used to set the value of the metric's datapoint. The column's values must be of an integer type. This may be case-sensitive, depending on the driver (i.e. Oracle DB).
```
I just noticed this sentence I added that is no longer true. Mind deleting it while you're in here?
```suggestion
* `value_column`(required): the column name in the returned dataset used to set the value of the metric's datapoint. This may be case-sensitive, depending on the driver (i.e. Oracle DB).
```

pmcollins replied:
Also, nit, but I believe `i.e.` is generally accepted to mean "that is" or "in other words". So I think you'll want this to be something like `e.g.`. Here and on line 27. 😄

crobert-1 replied:
Good catch, I'll update.

crobert-1 committed:
Fix misspelling issue causing lint to fail

pmcollins commented on receiver/sqlqueryreceiver/integration_test.go:
```diff
@@ -171,7 +171,10 @@ func testMovieMetrics(t *testing.T, rm pmetric.ResourceMetrics) {
 		pt := metric.Gauge().DataPoints().At(0)
-		genre, _ := pt.Attributes().Get("genre")
+		genre, exists := pt.Attributes().Get("genre")
+		if !exists {
```
What do you think about instead of this, passing in the name of the column, or perhaps an isOracle boolean?

crobert-1 replied:
Passing in the name of the key should work, as that seems more generic. The current layout of the test requires it to be something genre specific (since the variable is called genre), but in case other drivers end up being case-sensitive as well, it makes sense to me to pass in the genre attribute key.

crobert-1 committed:
Changes requested in PR

- Remove outdated README statement
- Correct i.e. in README to be e.g.
- Pass in variable in testing instead of special casing on failure

crobert-1 committed:
Update depenencies for failing test

dmitryax commented on Makefile.Common:
```diff
@@ -4,7 +4,7 @@ SRC_ROOT := $(realpath $(dir $(lastword $(MAKEFILE_LIST))))
 GO_BUILD_TAGS=""
 GOTEST_OPT?= -race -v -timeout 300s --tags=$(GO_BUILD_TAGS)
-GOTEST_INTEGRATION_OPT?= -race -timeout 120s
+GOTEST_INTEGRATION_OPT?= -race -timeout 900s
```
Did you add a test that runs 15 minutes?

crobert-1 replied:
Yes. This is to allow the Oracle DB docker image to be created, which takes about 10 minutes on a local machine. 

There is no published supported Oracle DB docker image, the supported method of creating docker images is to download a specific repository, and run a script. This is the `Setup Oracle DB Docker Image Dependency` step I added in the `build-and-test` workflow.

dmitryax replied:
What about this one https://container-registry.oracle.com/ords/f?p=113:4:114145384398230:::4:P4_REPOSITORY,AI_REPOSITORY,AI_REPOSITORY_NAME,P4_REPOSITORY_NAME,P4_EULA_ID,P4_BUSINESS_AREA_ID:8,8,Oracle%20Database%20Standard%20Edition%202,Oracle%20Database%20Standard%20Edition%202,1,0&cs=35MRG-gBMZGZd7cPIsIXl-GsL7HDV1AQqJDptQwn_G5F0noHF73yy9RusjZsCAmt1Zfmi3Cgq9sTMn5HXGtIuhw ?

dmitryax replied:
Otherwise please make sure that the image is cached on the CI. Additional 15 minutes to every CI run is a lot.

crobert-1 replied:
The link you included worked, that's a much nice way to handle this. Thanks for finding that, not sure how I didn't find it in my search!

I've updated as requested, and reverted the timeout to the original 120s.

dmitryax commented on receiver/sqlqueryreceiver/README.md:
````diff
@@ -83,3 +84,9 @@ Data point attributes:
 Value: 1
 ```
+
+#### Oracle DB Driver Example
+Refer to the config file [provided](./testdata/oracledb-receiver-config.yaml) for an example of using the
````
```suggestion
#### Oracle DB Driver Example

Refer to the config file [provided](./testdata/oracledb-receiver-config.yaml) for an example of using the
```

crobert-1 replied:
Fixed

pmcollins commented on receiver/sqlqueryreceiver/integration_test.go:
```diff
@@ -34,6 +34,83 @@ import (
 	"go.opentelemetry.io/collector/pdata/pmetric"
 )
 
+func testMovieMetrics(t *testing.T, rm pmetric.ResourceMetrics, genreAttrKey string) {
```
Not asking for this to be changed at all, but (since it looks like these functions got moved up) I tend to prefer putting functions below where they are called -- so a `main` function would be at the top, any functions it calls would be below it, any functions those call would be below them, etc.. This might be the opposite of the convention in Python 😄.

crobert-1 replied:
I've reordered the functions as you requested. I'm used to having helper methods declared and implemented before they're called, so that was why I originally moved things around. I've reverted to be more consistent with the rest of the repository.

crobert-1 committed:
Change requested

- Add empty line between header and content in README

crobert-1 committed:
Changes requested in PR

- Use container image cache instead of building image for
Oracle DB testing
- Update CHANGELOG

crobert-1 committed:
Add CHANGELOG issue reference

crobert-1 committed:
Change requested in PR

- Move test helper methods below tests

crobert-1 committed:
Add missing imports

dmitryax commented on .github/workflows/build-and-test.yml:
```diff
@@ -208,6 +208,10 @@ jobs:
             ~/go/bin
             ~/go/pkg/mod
           key: go-cache-${{ runner.os }}-${{ hashFiles('**/go.sum') }}
+      - name: Setup Oracle DB Docker Image Dependency
+        shell: bash
+        run: |
+          docker pull container-registry.oracle.com/database/express:latest
```
I don't think you need to explicitly pull the image. Integration tests should handle this

crobert-1 committed:
Change requested in PR

- Remove explicit pull as it shouldn't be required.

dmitryax commented on .github/workflows/build-and-test.yml:
```diff
@@ -208,7 +208,6 @@ jobs:
             ~/go/bin
             ~/go/pkg/mod
           key: go-cache-${{ runner.os }}-${{ hashFiles('**/go.sum') }}
-      - name: Run Integration Tests
```
oops :)

crobert-1 committed:
Undo accidental deleted line

dmitryax commented:
`TestOracleDBIntegration` is failing, PTAL

crobert-1 committed:
Revert CI changes

crobert-1 committed:
Remove github CI action again, increase timeout limit

dmitryax commented on Makefile.Common:
```diff
@@ -4,7 +4,7 @@ SRC_ROOT := $(realpath $(dir $(lastword $(MAKEFILE_LIST))))
 GO_BUILD_TAGS=""
 GOTEST_OPT?= -race -v -timeout 300s --tags=$(GO_BUILD_TAGS)
-GOTEST_INTEGRATION_OPT?= -race -timeout 120s
+GOTEST_INTEGRATION_OPT?= -race -timeout 300s
```
Can we try 3 minutes first?

crobert-1 replied:
I was expecting the test to fail at 5 minutes, but if it passes, I can drop it down to 3 minutes for another test run.

crobert-1 replied:
The successful test run took 289s, and the test failed with a 3 minute timeout. I think 5 minutes is a good number for now, but if we see failures with any frequency we can increase the timeout as well.

dmitryax replied:
Ok. Sounds good

crobert-1 committed:
Update timeout to be 3 minutes as requested

crobert-1 committed:
Change timeout to 5 minutes
